### DISCUSSION TITLE
NixOS support

### DISCUSSION THREAD

**Discussion by ghost** [IceWhaleTech/CasaOS#268] [Ideas]
What about CasaOS on NixOS?

**Comment by JohnGuan** [comment]
I've converted this to idea discussion, can you expand on that?

**Reply by ghost** [reply]
Because I want to learn it and I want to learn their configuration system. Since I can't contribute with some code that will install it on NixOS in the install script, I'm opening a issue.

**Comment by Trevo525** [comment]
I like this idea. CasaOS would be a nice web-interface for a NixOS server. However, CasaOS would have to manage a [nixpkg](https://nixos.org/manual/nixpkgs/stable/) rather than installing it from the shell script. 

NixOS is built with the principle that your entire system should be installed with a configuration file. All applications, users, and services are [specified in a file](https://nixos.org/manual/nixos/stable/#sec-changing-config). The actual system files are installed in a specific location and linked to the normal system locations. Like (/local/bin/)This allows you to reinstall by providing your config and typing `nixos-rebuild switch`. It also attempts to solve [dependency hell](https://en.wikipedia.org/wiki/Dependency_hell) because you can have many versions of the same library and it each application is linked to the packaged dependencies. Also, if you do an update and something breaks, you can go to previous versions with a reboot and it will swap out the links to the applications of that "build". It is a bit of a learning curve and it is a bit more wasteful on space. But, reinstalling, updates, and changes to the OS  are made very safe and simple. 

Other than the extra work of maintaining a nixpkg, the only other issue I  see is the docker installation. CasaOS would have to let NixOS handle that. You would pretty much just have to point the user to the [docs](https://nixos.wiki/wiki/Docker). But, if you wanted to use NixOS, that would be the users' problem to deal with.
